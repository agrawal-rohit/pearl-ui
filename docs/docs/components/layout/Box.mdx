---
sidebar_position: 1
title: Box
---

import SourceButton from "../../../src/components/SourceButton/SourceButton";

<div style={{ display: "flex" }}>
  <SourceButton
    label="Source"
    href="https://github.com/agrawal-rohit/pearl-ui/blob/main/src/components/atoms/box/box.tsx"
  />
</div>

The `Box` component is a fundamental building block in the Pearl UI library. It is a highly abstract component that serves as the foundation for all other components in Pearl UI. By default, it renders a [React Native View](https://reactnative.dev/docs/view) element.

The Box component enhances the development experience by providing the following benefits

1. **Simplified Layout Creation**: The Box component allows developers to create flexible layouts with ease.
2. **Style Prop Support**: It provides a more intuitive way to pass styles via props. For instance, `backgroundColor` can be passed as a prop instead of using [React Stylesheet](https://reactjs.org/docs/faq-styling.html) or inline styles. Shorthand props are also supported (e.g., `bgColor` instead of `backgroundColor`).
3. **Base for Complex Components**: The Box component can be used as the base component for creating more complex components, replacing the need for `View`. All components in Pearl UI use `Box` as their base.

## Import

```jsx
import { Box } from "pearl-ui";
```

## Usage

```jsx
<Box width="100%" height={200} backgroundColor="primary.500" marginBottom="2">
  This is a box
</Box>
```

## Example

import Snack from "../../../src/components/ExpoSnack";

<Snack snackId="@agrawal-rohit/box" />

## Component Properties

### Supported Style Properties

The `Box` component supports the following style props:

- [backgroundColor](../../overview/core-features/style-props#color-and-background-color) (Note: Only `backgroundColor` props are supported, not the `color` prop)
- [layout](../../overview/core-features/style-props#layout)
- [transform](../../overview/core-features/style-props#transform)
- [position](../../overview/core-features/style-props#position)
- [opacity and visibility](../../overview/core-features/style-props#opacity-and-visibility)
- [margin and padding](../../overview/core-features/style-props#margin-and-padding)
- [border](../../overview/core-features/style-props#border)
- [shadow](../../overview/core-features/style-props#shadow)
- [shadow](../../overview/core-features/style-props#shadow)

### Animation Properties

The `Box` component also supports animation properties from the [Moti](https://moti.fyi/) library, which allows for the creation of dynamic and interactive UI elements. The [useMotiWithStyleProps](../../utils/hooks/useMotiWithStyleProps) hook from the Moti library is utilized to manage these animation properties.

| Property              | Type                        | Default     | Description                                                                                                                                                                                                                                                                 |
| --------------------- | --------------------------- | ----------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `from`                | <t>object</t>               |             | Defines the starting state of the animation.                                                                                                                                                                                                                                |
| `to`                  | <t>object</t>               |             | Defines the ending state of the animation.                                                                                                                                                                                                                                  |
| `transition`          | <t>object</t>               |             | Specifies the type of transition that the animation will use.                                                                                                                                                                                                               |
| `exitTransition`      | <t>object</t>               |             | Determines the transition type that will be used when the component is unmounted. By default, `exit` uses `transition` to configure its animations, so this prop is not required. However, if you pass `exitTransition`, it will override `transition` for exit animations. |
| `state`               | <t>object</t>               |             | Manages the overall state of the animation, including the `from` and `to` states.                                                                                                                                                                                           |
| `delay`               | <t>number</t>               |             | Manage the delay for the `animate` field. (Use the `transition` prop for more granular control)                                                                                                                                                                             |
| `onDidAnimate`        | <t>() =&gt; void</t>        |             | Callback function called after finishing an animation.                                                                                                                                                                                                                      |
| `stylePriority`       | <t>"state" \| "animate"</t> | `"animate"` | This is not a prop you will likely find yourself using. If set to <t>"animate"</t>, then styles passed from the `animate` prop will take precedent. Otherwise, if set to <t>"state"</t>, then the `state` prop will take precedent for matching styles.                     |
| `animateInitialState` | <t>boolean</t>              | `false`     | If `true`, the `from` prop will be set to animate. This will be noticeable for some spring animations, and might make them jumpy on mount.                                                                                                                                  |
