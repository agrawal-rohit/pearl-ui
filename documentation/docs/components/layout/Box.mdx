---
sidebar_position: 1
title: Box
---

import SourceButton from "../../../src/components/SourceButton/SourceButton";

<div style={{ display: "flex" }}>
  <SourceButton
    label="Source"
    href="https://github.com/agrawal-rohit/pearl-ui/blob/main/src/components/atoms/box/box.tsx"
  />
</div>

The `Box` component is a fundamental building block in the Pearl UI library. It is a highly abstract component that serves as the foundation for all other components in Pearl UI. By default, it renders a [React Native View](https://reactnative.dev/docs/view) element.

The Box component enhances the development experience by providing the following benefits

1. **Simplified Layout Creation**: The Box component allows developers to create flexible layouts with ease.
2. **Style Prop Support**: It provides a more intuitive way to pass styles via props. For instance, `backgroundColor` can be passed as a prop instead of using [React Stylesheet](https://reactjs.org/docs/faq-styling.html) or inline styles. Shorthand props are also supported (e.g., `bgColor` instead of `backgroundColor`).
3. **Base for Complex Components**: The Box component can be used as the base component for creating more complex components, replacing the need for `View`. All components in Pearl UI use `Box` as their base.

## Import

```jsx
import { Box } from "pearl-ui";
```

## Usage

```jsx
<Box width="100%" height={200} backgroundColor="primary.500" marginBottom="m">
  This is a box
</Box>
```

## Example

import Snack from "../../../src/components/ExpoSnack";

<Snack snackId="@agrawal-rohit/box" />

## Component Properties

### Supported Style Properties

The `Box` component supports the following style props:

- [backgroundColor](../../core-features/style-props#color-and-background-color) (Note: Only `backgroundColor` props are supported, not the `color` prop)
- [layout](../../core-features/style-props#layout)
- [position](../../core-features/style-props#position)
- [opacity and visibility](../../core-features/style-props#opacity-and-visibility)
- [margin and padding](../../core-features/style-props#margin-and-padding)
- [border](../../core-features/style-props#border)
- [shadow](../../core-features/style-props#shadow)
- [shadow](../../core-features/style-props#shadow)

### Animation Properties

The `Box` component also supports animation properties from the [Moti](https://moti.fyi/) library, which allows for the creation of dynamic and interactive UI elements. The [useMotiWithStyleProps](../../hooks/useMotiWithStyleProps) hook from the Moti library is utilized to manage these animation properties.

| Property         | Type          | Description                                                                       |
| ---------------- | ------------- | --------------------------------------------------------------------------------- |
| `from`           | <t>object</t> | Defines the starting state of the animation.                                      |
| `to`             | <t>object</t> | Defines the ending state of the animation.                                        |
| `transition`     | <t>object</t> | Specifies the type of transition that the animation will use.                     |
| `exitTransition` | <t>object</t> | Determines the transition type that will be used when the component is unmounted. |
| `state`          | <t>object</t> | Manages the overall state of the animation, including the `from` and `to` states. |
