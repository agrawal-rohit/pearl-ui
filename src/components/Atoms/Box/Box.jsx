import React from "react";
import { View } from "react-native";
import useStyledProps from "../../../hooks/useStyledProps";
import { useTheme } from "../../../hooks/useTheme";
import {
  backgroundColor,
  border,
  layout,
  opacity,
  position,
  shadow,
  spacing,
  visible,
} from "../../../theme/src/styleFunctions";
import {
  IBasePearlTheme,
  RestyleFunctionContainer,
} from "../../../theme/src/types";

// export type BoxProps<
//   Theme extends IBasePearlTheme
// > = BackgroundColorProps<Theme> &
//   OpacityProps &
//   VisibleProps &
//   LayoutProps &
//   SpacingProps<Theme> &
//   BorderProps<Theme> &
//   ShadowProps<Theme> &
//   PositionProps<Theme>;

export const boxRestyleFunctions = [
  backgroundColor,
  opacity,
  visible,
  layout,
  spacing,
  border,
  shadow,
  position,
];

/**
 * Box is the most abstract component on top of which all other Pearl UI components are built. By default, it renders a <View> element.
 */
// type ViewProps = React.ComponentProps<typeof View> & {
//   children?: React.ReactNode;
// };
// type ComponentProps<Theme extends IBasePearlTheme> = BoxProps<Theme> &
//   Omit<ViewProps, keyof BoxProps<Theme>>;

const Box = React.forwardRef((props, ref) => {
  const passedProps = useStyledProps(boxRestyleFunctions, props);

  return <View ref={ref} {...passedProps} />;
});

export default Box;
