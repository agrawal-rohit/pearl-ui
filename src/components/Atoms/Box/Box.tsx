import React from "react";
import { View, ViewProps } from "react-native";
import { pearlify } from "../../../pearlify";

/**
 * Box is the most abstract component on top of which all other Pearl UI components are built.
 * By default, it renders a <View> element.
 */
const Box = pearlify<ViewProps>(View, {
  componentName: "Box",
  type: "basic",
  animatable: true,
});

// The props that the Box component accepts.
export type BoxProps = React.ComponentProps<typeof Box>;

// Export the Box component as the default export.
export default Box;
