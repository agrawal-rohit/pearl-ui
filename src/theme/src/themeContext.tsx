import AppLoading from "expo-app-loading";
import {
  useFonts,
  Poppins_300Light,
  Poppins_400Regular,
  Poppins_500Medium,
  Poppins_600SemiBold,
  Poppins_700Bold,
  Poppins_800ExtraBold,
} from "@expo-google-fonts/poppins";
import React, { useEffect, useState, createContext } from "react";
import { useColorScheme } from "react-native";
import { baseTheme } from "./basetheme";
import { BasePearlTheme } from "./types";

type ThemeType = "light" | "dark" | "system";

export interface IThemeContext {
  /** Theme configuration object for the active color mode */
  theme: BasePearlTheme;
  /** Active color mode */
  colorMode: ThemeType;
  /** Function to toggle the active color mode */
  toggleColorMode(): void;
}

interface ThemeProviderProps {
  /** Initial color mode for the theme (light, dark, system) */
  initialColorMode?: ThemeType;
  /** The theme configuration object */
  theme?: BasePearlTheme;
  /** A mapping object of the fonts used by the theme */
  fonts?: object;
  /**React children */
  children: React.ReactNode;
}

const baseFonts = {
  "Poppins-Light": Poppins_300Light,
  "Poppins-Regular": Poppins_400Regular,
  "Poppins-Medium": Poppins_500Medium,
  "Poppins-SemiBold": Poppins_600SemiBold,
  "Poppins-Bold": Poppins_700Bold,
  "Poppins-ExtraBold": Poppins_800ExtraBold,
};

export const themeContext = createContext({} as IThemeContext);

/**
 * The main provider component which controls the theme and color mode of the components in the entire application.
 */
export const ThemeProvider: React.FC<ThemeProviderProps> = ({
  initialColorMode = "light",
  theme = baseTheme,
  fonts = baseFonts,
  children,
}) => {
  const [colorMode, setColorMode] = useState<ThemeType>(initialColorMode);
  const systemThemeStyle = useColorScheme() as ThemeType;
  const [haveFontsLoaded] = useFonts(fonts);

  const changeThemeTo = (colorMode: ThemeType) => {
    if (colorMode === "dark") {
      setColorMode("dark");
    } else {
      setColorMode("light");
    }
  };

  const toggleColorMode = () => {
    if (colorMode === "light") {
      changeThemeTo("dark");
    } else {
      changeThemeTo("light");
    }
  };

  useEffect(() => {
    // Set initial Theme
    if (colorMode === "light") {
      changeThemeTo("light");
    } else if (colorMode === "dark") {
      changeThemeTo("dark");
    } else if (colorMode === "system") {
      changeThemeTo(systemThemeStyle);
    }
  }, [colorMode, initialColorMode, theme]);

  if (haveFontsLoaded) {
    return (
      <themeContext.Provider
        value={{ theme, colorMode, toggleColorMode } as IThemeContext}
      >
        {children}
      </themeContext.Provider>
    );
  }

  return <AppLoading />;
};
