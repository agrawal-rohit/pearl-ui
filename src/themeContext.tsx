import React, { useEffect, useState, useContext, createContext } from "react";
import { useColorScheme } from "react-native";
import { ThemeProvider as RestyleThemeProvider } from "@shopify/restyle";
import { baseLightTheme, baseDarkTheme, Theme } from "./theme";

type ThemeType = "light" | "dark" | "system";

interface IThemeContext {
  /** Theme configuration object for the active color mode */
  theme: Theme;
  /** Active color mode */
  colorMode: ThemeType;
  /** Function to toggle the active color mode and corresponding theme */
  toggleTheme(theme: string): void;
}

interface ThemeProviderProps {
  /** Default color mode for the theme */
  defaultColorMode?: ThemeType;
  /** The configuration object for light theme */
  lightTheme?: Theme;
  /** The configuration object for dark theme */
  darkTheme?: Theme;
}

export const themeContext = createContext({} as IThemeContext);

/**
 * The main provider component which controls the theme and color mode of the components in the entire application.
 */
export const ThemeProvider: React.FC<ThemeProviderProps> = ({
  defaultColorMode = "light",
  lightTheme = baseLightTheme,
  darkTheme = baseDarkTheme,
  children,
}) => {
  const [activeTheme, setActiveTheme] = useState<Theme>(lightTheme);
  const [colorMode, setColorMode] = useState<ThemeType>(defaultColorMode);
  const systemThemeStyle = useColorScheme() as ThemeType;

  const toggleTheme = () => {
    if (colorMode === "light") {
      setColorMode("dark");
      setActiveTheme(darkTheme);
    } else {
      setColorMode("light");
      setActiveTheme(lightTheme);
    }
  };

  useEffect(() => {
    // Set theme
    if (defaultColorMode === "system") {
      setColorMode(systemThemeStyle);
      setActiveTheme(systemThemeStyle === "light" ? lightTheme : darkTheme);
    }
  }, []);

  return (
    <themeContext.Provider
      value={{ theme: activeTheme, colorMode, toggleTheme } as IThemeContext}
    >
      <RestyleThemeProvider theme={activeTheme}>
        {children}
      </RestyleThemeProvider>
    </themeContext.Provider>
  );
};

/**
 * Hook to get access to the active theme object, active color mode, and a function to toggle the active theme
 */
export const useTheme = () => useContext(themeContext);
