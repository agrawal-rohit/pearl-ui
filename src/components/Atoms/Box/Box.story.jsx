import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "../Screen/Screen";
import Box from "./Box";

storiesOf("Box", module)
  .addDecorator((getStory) => getStory())
  .add("Basic box", () => (
    <Box
      padding="s"
      height={300}
      width={100}
      backgroundColor="primary100"
    ></Box>
  ));
