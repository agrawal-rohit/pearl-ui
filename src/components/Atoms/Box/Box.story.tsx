import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "../Screen/Screen";
import Box from "../Box/Box";

storiesOf("Box", module)
  .addDecorator((getStory) => <Screen>{getStory()}</Screen>)
  .add("Basic box", () => (
    <Box padding="s" width={200} backgroundColor="primary100"></Box>
  ));
