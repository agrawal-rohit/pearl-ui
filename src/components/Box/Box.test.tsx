import React from "react";
import Box from "./Box";
import { ThemeProvider } from "../../themeContext";
import { render } from "@testing-library/react-native";

jest.useFakeTimers();

describe("<Box />", () => {
  it("exists", () => {
    const tree = render(
      <ThemeProvider>
        <Box>asdasd</Box>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
