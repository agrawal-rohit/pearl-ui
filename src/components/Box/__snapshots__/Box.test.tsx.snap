// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Box /> exists 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
      },
    ]
  }
>
  asdasd
</View>
`;
